packages/greedy_infomax/src/nupic/research/frameworks/greedy_infomax/models/gim_block.py

```diff
@@ -0,0 +1,162 @@
+                 k_predictions=5,
+                 n_patches_x=None,
+                 n_patches_y=None,):
+        """
+        A block that can be placed after any module in a model which consists of:
+        1. A BilinearInfo module
+        2. An EmitEncoding module
+        3. A GradientBlock module
+
+        In GreedyInfoMax experiments, this block represents the segregation of the
+        gradients between the modules that come before and the modules that come
+        after this block. These can be placed after any module in a model. Note that
+        it is not completely necessary that the gradients be blocked, or even that
+        the EmitEncoding blocks to be placed after the BilinearInfo block, but this
+        is the most common use case.
+        """
+        super(GreedyInfoMaxBlock, self).__init__()
+        self.info_estimate_aggregator = estimate_info_aggregator
+        self.encoding_aggregator = encoding_aggregator
+        self.negative_samples = negative_samples
+        self.k_predictions = k_predictions
+        self.in_channels = in_channels
+        self.bilinear_info = BilinearInfo(in_channels,
+                                          in_channels,
+                                          self.negative_samples,
+                                          self.k_predictions)
+        self.emit_encoding = EmitEncoding(in_channels)
+        self.gradient_block = GradientBlock()
+        self.n_patches_x, self.n_patches_y = n_patches_x, n_patches_y
+
+    def forward(self, x):
+        out = F.adaptive_avg_pool2d(x, 1)
+        out = out.reshape(-1, self.n_patches_x, self.n_patches_y, out.shape[1])
+        out = out.permute(0, 3, 1, 2).contiguous()
+        info_estimate = self.bilinear_info.estimate_info(out, out)
+        self.info_estimate_aggregator.append(info_estimate)
+        x_blocked = self.gradient_block(x).clone()
+        return x_blocked
+
+    """
+    During unsupervised training, this function will be linked to the forward hook
+    for its corresponding module.
+    """
+    def wrapped_forward(self, module, input, output):  # noqa: A002
+        return self.forward(output)
+
+    def encode(self, x):
+        encoded = self.emit_encoding.encode(x, self.n_patches_x, self.n_patches_y)
+        self.encoding_aggregator.append(encoded)
+
+    """
+    During supervised training, this function will be linked to the forward hook for
+    its corresponding module.
+    """
+    def wrapped_encode(self, module, input, output):  # noqa: A002
+        return self.encode(output)
+
+
+class SparseGreedyInfoMaxBlock(GreedyInfoMaxBlock):
+    """
+    A version of the above GreedyInfoMaxBlock which uses SparseBilinearInfo instead
+    of a regular BilinearInfo module.
+    """
+    def __init__(self,
+                 estimator_outputs,
+                 encoding_outputs,
+                 in_channels,
+                 negative_samples=16,
+                 k_predictions=5,
+                 sparsity=None):
```

Comment:
`sparsity` shouldn't be None, it should default to some numeric value instead.

Reply:
Now defaults to 0.0